### PR TITLE
Introduce ICA Keeper (deduplicating host/controller code)

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

Introduces ICAKeeper to deduplicate code within controller and host. The controller and host modules will be given a prefixed store from the ICAKeeper. Chains not wishing to run controller or host code do not need to create the respective keepers for those modules

This code allows for multiple controller sub modules to be created on the same chain!

closes: #563 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go [hunk 1]
```diff
@@ -93,91 +69,52 @@ func (k Keeper) ClaimCapability(ctx sdk.Context, cap *capabilitytypes.Capability
 	return k.scopedKeeper.ClaimCapability(ctx, cap, name)
 }
 
-// GetActiveChannelID retrieves the active channelID from the store keyed by the provided portID
-func (k Keeper) GetActiveChannelID(ctx sdk.Context, portID string) (string, bool) {
-	store := ctx.KVStore(k.storeKey)
-	key := types.KeyActiveChannel(portID)
+// GetAllPorts returns all ports to which the interchain accounts controller module is bound. Used in ExportGenesis
+func (k Keeper) GetAllPorts(ctx sdk.Context) []string {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper.go:73]: All duplicate keeper funcs in controller/host are now wrappers

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go [hunk 2]
```diff
@@ -32,17 +32,17 @@ type Keeper struct {
...
-	cdc codec.BinaryCodec, key sdk.StoreKey,
+	cdc codec.BinaryCodec, storePrefix string, icaKeeper icakeeper.Keeper,
 	ics4Wrapper types.ICS4Wrapper, channelKeeper types.ChannelKeeper, portKeeper types.PortKeeper,
-	accountKeeper types.AccountKeeper, scopedKeeper capabilitykeeper.ScopedKeeper, msgRouter *baseapp.MsgServiceRouter,
+	scopedKeeper capabilitykeeper.ScopedKeeper, msgRouter *baseapp.MsgServiceRouter,
 ) Keeper {
 	return Keeper{
-		storeKey:      key,
 		cdc:           cdc,
+		storePrefix:   storePrefix,
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper.go:41]: prefix used to create a prefixed store from the ICAKeeper

So, one store, but each ica submodule has its own namespace. This is actual quite useful. Now we can create many controller modules (currently require a new keeper for each)

File: modules/apps/27-interchain-accounts/keeper/keeper.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+// prefixStore returns isolated prefix store for each ICA submodule so they can read/write in separate
+// namespace without being able to read/write to each other's data.
+func (k Keeper) prefixStore(ctx sdk.Context, storePrefix string) sdk.KVStore {
+	return prefix.NewStore(ctx.KVStore(k.storeKey), []byte(storePrefix))
+}
+
+// GetAllPorts returns all ports to which a interchain accounts submodule is bound. Used in ExportGenesis
+func (k Keeper) GetAllPorts(ctx sdk.Context, storePrefix string) []string {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/keeper/keeper.go:48]: I didn't change any logic in these functions except the first line. Replacing the ctx.KVStore(k.storeKey) with the prefixStore call

File: modules/apps/27-interchain-accounts/keeper/keeper_test.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
+package keeper_test
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/keeper/keeper_test.go:1]: This test file is copied from controller keeper. Didn't change any logic except calling ICAKeeper directly with icatypes.ControllerModuleName

**Review by colin-axner — COMMENTED** [review]
I think this is a nice improvement and cuts down on technical debt.

There's one more change necessary to enable support for many ica auth modules (each namespaced). To remove the pointer in the InitGenesis/ExportGenesis logic, we should replace the individual host/controller init/export with a single function at the ica keeper level. The genesis should contain a repeated list of ica submodule genesis, each of which contains the existing fields + the store prefix

My main concern about the store prefix is name collisions. We might want to have some sort of registration function to ensure there are no name collisions. Thoughts?

**Comment by colin-axner** [comment]
After some more reflection, I worry a little too much about introducing reliance on store prefixes not colliding. Chains should be able to register multiple controller modules since they can create multiple IBC controller modules and the keeper functions all use portID for mappings, thus owner addresses could be equivalent across controller modules (assuming we modify the portID to namespace channelID as well). 

Thus, the main benefit would be code deduplication. I don't think this trade off is worth it right now given that the SDK is making significant changes to how the stores will work. Once the SDK moves to SMT, we should revisit this discussion. Closing this pr and I will update the issue with my thoughts

**Comment by seantking** [comment]
Thanks for exploring this regardless. I learned a few things :)

**Pull request closed without merging**